src/backend/session/libseat.rs

```diff
@@ -0,0 +1,261 @@
+//!
+//! Implementation of the [`Session`](::backend::session::Session) trait through the libseat.
+//!
+//! This requires libseat to be available on the system.
+
+use libseat::Seat;
+use std::{
+    cell::RefCell,
+    collections::HashMap,
+    os::unix::io::RawFd,
+    path::Path,
+    rc::{Rc, Weak},
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+};
+
+use nix::{fcntl::OFlag, unistd::close};
+
+use calloop::{EventSource, Poll, Readiness, Token};
+
+use crate::{
+    backend::session::{AsErrno, Session, Signal as SessionSignal},
+    signaling::Signaler,
+};
+
+#[derive(Debug)]
+struct LibSeatSessionImpl {
+    seat: RefCell<Seat>,
+    active: Arc<AtomicBool>,
+    devices: RefCell<HashMap<RawFd, i32>>,
+    logger: ::slog::Logger,
+}
+
+impl Drop for LibSeatSessionImpl {
+    fn drop(&mut self) {
+        debug!(self.logger, "Closing seat")
+    }
+}
+
+/// [`Session`] via the libseat
+#[derive(Debug, Clone)]
+pub struct LibSeatSession {
+    internal: Weak<LibSeatSessionImpl>,
+    seat_name: String,
+}
+
+/// [`SessionNotifier`] via the libseat
+#[derive(Debug)]
+pub struct LibSeatSessionNotifier {
+    internal: Rc<LibSeatSessionImpl>,
+    signaler: Signaler<SessionSignal>,
+}
+
+impl LibSeatSession {
+    /// Tries to create a new session via libseat.
+    pub fn new<L>(logger: L) -> Result<(LibSeatSession, LibSeatSessionNotifier), Error>
+    where
+        L: Into<Option<::slog::Logger>>,
+    {
+        let logger = crate::slog_or_fallback(logger)
+            .new(o!("smithay_module" => "backend_session", "session_type" => "libseat"));
+
+        let active = Arc::new(AtomicBool::new(false));
+        let signaler = Signaler::new();
+
+        let seat = {
+            let enable = {
+                let active = active.clone();
+                let signaler = signaler.clone();
+                let logger = logger.clone();
+                move |_seat: &mut libseat::SeatRef| {
+                    debug!(logger, "Enable callback called");
+                    active.store(true, Ordering::SeqCst);
+                    signaler.signal(SessionSignal::ActivateSession);
+                }
+            };
+            let disable = {
+                let active = active.clone();
+                let signaler = signaler.clone();
+                let logger = logger.clone();
+                move |seat: &mut libseat::SeatRef| {
+                    debug!(logger, "Disable callback called");
+                    active.store(false, Ordering::SeqCst);
+                    seat.disable().unwrap();
+                    signaler.signal(SessionSignal::PauseSession);
+                }
+            };
+
+            Seat::open(enable, disable)
```

Comment:
This is nitpicking, but would it make sense to have, in the API of your libseat bindings, this init function just take a single closure and pass it an enum argument to distinguish enable/disable events?

That could help with reducing code duplication a little.

Reply:
Sure, I don't like it either, so I'm happy to change that. I will work on that tomorrow

Reply:
Done https://github.com/Smithay/smithay/pull/292/commits/7b746483e43e6ae859449a4d1ae70a369f824036